README
NAME
     Tags::Output::PYX - PYX class for line oriented output for 'Tags'.

SYNOPSYS
     use Tags::Output::PYX;

     my $obj = Tags::Output::PYX->new(%parameters);
     $obj->finalize;
     my $ret = $obj->flush($reset_flag);
     my @elements = $obj->open_elements;
     $obj->put(@data);
     $obj->reset;

     # Deprecated methods.
     my @elements = $obj->open_tags;

PYX LINE CHARS
     ?  - Instruction.
     (  - Open tag.
     )  - Close tag.
     A  - Attribute.
     -  - Normal data.

METHODS
  "new"
     my $obj = Tags::Output::PYX->new(%parameters);

    Constructor.

    Returns instance of object.

    *       "auto_flush"

             Auto flush flag.
             Default value is 0.

    *       "output_callback"

             Output callback.
             Default value is undef.

    *       "output_handler"

             Set output handler.
             Default value is undef.

    *       "output_sep"

             Output separator.
             Default value is newline.

    *       "skip_bad_data"

             Skip bad tags.
             Default value is 0.

    *       "strict_instruction"

             Strict instruction.
             Default value is 1.

  "finalize"
     $obj->finalize;

    Finalize Tags output. Automaticly puts end of all opened tags.

    Returns undef.

  "flush"
     my $ret = $obj->flush($reset_flag);

    Flush tags in object. If defined 'output_handler' flush to its. Or
    returns code. If enabled $reset_flag, then resets internal variables via
    reset method.

  "open_elements"
     my @elements = $obj->open_elements;

    Get list of open elements.

    Returns array of strings.

  "put"
     $obj->put(@data);

    Put tags code in tags format.

    Returns undef.

  "reset"
     $obj->reset;

    Resets internal variables.

    Returns undef.

DEPRECATED METHODS
  "open_tags"
     my @elements = $obj->open_tags;

    Get list of open elements.

    Returns array of strings.

ERRORS
     new():
             Auto-flush can't use without output handler.
             Output handler is bad file handler.
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

     flush():
             Cannot write to output handler.

     put():
             Bad data.
             Bad type of data.
             Bad number of arguments. 'Tags' structure %s 
             Ending bad tag: '%s' in block of tag '%s'.

EXAMPLE
     use strict;
     use warnings;

     use Tags::Output::PYX;

     # Object.
     my $tags = Tags::Output::PYX->new;

     # Put all tag types.
     $tags->put(
             ['b', 'tag'],
             ['a', 'par', 'val'],
             ['c', 'data', \'data'],
             ['e', 'tag'],
             ['i', 'target', 'data'],
             ['b', 'tag'],
             ['d', 'data', 'data'],
             ['e', 'tag'],
     );

     # Print out.
     print $tags->flush."\n";

     # Output:
     # (tag
     # Apar val
     # -<!--data--><!--SCALAR(0x1570740)-->
     # )tag
     # ?target data
     # (tag
     # -datadata
     # )tag

DEPENDENCIES
    Error::Pure, Readonly, Tags::Output, Tags::Utils.

SEE ALSO
    Tags
        Structure oriented SGML/XML/HTML/etc. elements manipulation.

    Task::PYX
        Install the PYX modules.

    Task::Tags
        Install the Tags modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/Tags-Output-PYX>

AUTHOR
    Michal Josef Špaček skim@cpan.org

LICENSE AND COPYRIGHT
    © 2011-2020 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.06



Changes
0.06

0.05 2020-05-16T20:10:35+02:00
 - Add new open_elements() method to doc.
 - Add test messages to tests.
 - Fix testing with both version of Tags::Output (<0.09 and >=0.09)
   RT#132534.
 - Rewrite doc to use =head2 for methods.

0.04 2020-05-11T15:37:49+02:00
 - Add REPOSITORY section to doc.
 - Fix Makefile.PL fails with no '.' in @INC.
 - Fix warnings in tests caused in new version of Tags::Output (RT#132534).
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.03 2015-10-15T22:03:52+02:00
 - Clean Error::Pure structure after error in tests.
 - Improve SEE ALSO section in doc.
 - Move author tests to xt/ directory.
 - Rewrite modules in SEE ALSO section to proper Tasks.
 - Sort tests.
 - Update Module::Install to 1.16 version.

0.02 2015-03-19T11:02:24+01:00
 - Add comments to tests.
 - Add missing tests for 'Test::NoWarnings'.
 - Rename tests for put to form ??-put-precision.t .
 - Update LICENSE AND COPYRIGHT section in doc to my standard.
 - Update copyright years.

0.01 2014-11-05T07:40:20+01:00
 - First version.


PYX.pm
package Tags::Output::PYX;

use base qw(Tags::Output);
use strict;
use warnings;

use Error::Pure qw(err);
use Readonly;
use Tags::Utils qw(encode_newline);

# Constants.
Readonly::Scalar my $EMPTY_STR => q{};
Readonly::Scalar my $SPACE => q{ };

our $VERSION = 0.06;

# Attributes.
sub _put_attribute {
	my ($self, $attr, $value) = @_;
	push @{$self->{'flush_code'}}, "A$attr $value";
	return;
}

# Begin of tag.
sub _put_begin_of_tag {
	my ($self, $tag) = @_;
	push @{$self->{'flush_code'}}, "($tag";
	unshift @{$self->{'printed_tags'}}, $tag;
	return;
}

# CData.
sub _put_cdata {
	my ($self, @cdata) = @_;
	$self->_put_data(@cdata);
	return;
}

# Comment.
sub _put_comment {
	my ($self, @comments) = @_;
	$self->_put_data(
		map { '<!--'.$_.'-->' } @comments,
	);
	return;
}

# Data.
sub _put_data {
	my ($self, @data) = @_;
	my $data = join($EMPTY_STR, @data);
	push @{$self->{'flush_code'}}, '-'.encode_newline($data);
	return;
}

# End of tag.
sub _put_end_of_tag {
	my ($self, $tag) = @_;
	my $printed = shift @{$self->{'printed_tags'}};
	if ($printed ne $tag) {
		err "Ending bad tag: '$tag' in block of tag '$printed'.";
	}
	push @{$self->{'flush_code'}}, ")$tag";
	return;
}

# Instruction.
sub _put_instruction {
	my ($self, $target, $code) = @_;

	# Construct instruction line.
	my $instruction = '?'.$target;
	if ($code) {
		$instruction .= $SPACE.$code;
	}

	# To flush code.
	push @{$self->{'flush_code'}}, encode_newline($instruction);

	return;
}

# Raw data.
sub _put_raw {
	my ($self, @raw_data) = @_;
	$self->_put_data(@raw_data);
	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

 Tags::Output::PYX - PYX class for line oriented output for 'Tags'.

=head1 SYNOPSYS

 use Tags::Output::PYX;

 my $obj = Tags::Output::PYX->new(%parameters);
 $obj->finalize;
 my $ret = $obj->flush($reset_flag);
 my @elements = $obj->open_elements;
 $obj->put(@data);
 $obj->reset;

 # Deprecated methods.
 my @elements = $obj->open_tags;

=head1 PYX LINE CHARS

 ?  - Instruction.
 (  - Open tag.
 )  - Close tag.
 A  - Attribute.
 -  - Normal data.

=head1 METHODS

=head2 C<new>

 my $obj = Tags::Output::PYX->new(%parameters);

Constructor.

Returns instance of object.

=over 8

=item * C<auto_flush>

 Auto flush flag.
 Default value is 0.

=item * C<output_callback>

 Output callback.
 Default value is undef.

=item * C<output_handler>

 Set output handler.
 Default value is undef.

=item * C<output_sep>

 Output separator.
 Default value is newline.

=item * C<skip_bad_data>

 Skip bad tags.
 Default value is 0.

=item * C<strict_instruction>

 Strict instruction.
 Default value is 1.

=back

=head2 C<finalize>

 $obj->finalize;

Finalize Tags output.
Automaticly puts end of all opened tags.

Returns undef.

=head2 C<flush>

 my $ret = $obj->flush($reset_flag);

Flush tags in object.
If defined 'output_handler' flush to its.
Or returns code.
If enabled $reset_flag, then resets internal variables via reset method.

=head2 C<open_elements>

 my @elements = $obj->open_elements;

Get list of open elements.

Returns array of strings.

=head2 C<put>

 $obj->put(@data);

Put tags code in tags format.

Returns undef.

=head2 C<reset>

 $obj->reset;

Resets internal variables.

Returns undef.

=head1 DEPRECATED METHODS

=head2 C<open_tags>

 my @elements = $obj->open_tags;

Get list of open elements.

Returns array of strings.

=head1 ERRORS

 new():
         Auto-flush can't use without output handler.
         Output handler is bad file handler.
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

 flush():
         Cannot write to output handler.

 put():
         Bad data.
         Bad type of data.
         Bad number of arguments. 'Tags' structure %s 
         Ending bad tag: '%s' in block of tag '%s'.

=head1 EXAMPLE

 use strict;
 use warnings;

 use Tags::Output::PYX;

 # Object.
 my $tags = Tags::Output::PYX->new;

 # Put all tag types.
 $tags->put(
         ['b', 'tag'],
         ['a', 'par', 'val'],
         ['c', 'data', \'data'],
         ['e', 'tag'],
         ['i', 'target', 'data'],
         ['b', 'tag'],
         ['d', 'data', 'data'],
         ['e', 'tag'],
 );

 # Print out.
 print $tags->flush."\n";

 # Output:
 # (tag
 # Apar val
 # -<!--data--><!--SCALAR(0x1570740)-->
 # )tag
 # ?target data
 # (tag
 # -datadata
 # )tag

=head1 DEPENDENCIES

L<Error::Pure>,
L<Readonly>,
L<Tags::Output>,
L<Tags::Utils>.

=head1 SEE ALSO

=over

=item L<Tags>

Structure oriented SGML/XML/HTML/etc. elements manipulation.

=item L<Task::PYX>

Install the PYX modules.

=item L<Task::Tags>

Install the Tags modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Tags-Output-PYX>

=head1 AUTHOR

Michal Josef Špaček L<skim@cpan.org>

=head1 LICENSE AND COPYRIGHT

© 2011-2020 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.06

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract "PYX class for line oriented output for 'Tags'.";
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Tags-Output-PYX';
readme_from 'PYX.pm';
recursive_author_tests('xt');
requires 'Error::Pure' => 0;
requires 'Readonly' => 0;
requires 'Tags::Output' => 0.02;
requires 'Tags::Utils' => 0;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://rt.cpan.org/Public/Dist/Display.html'.
	'?Name=Tags-Output-PYX';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Tags-Output-PYX';
resources 'repository' => 'git://github.com/michal-josef-spacek/Tags-Output-PYX';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.22;
test_requires 'File::Object' => 0.05;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
test_requires 'Test::Output' => 1.02;
test_requires 'Test::Warn' => 0;
test_requires 'version' => 0.77;
tests_recursive;
version '0.06';

# Run.
WriteAll();
